### PR TITLE
feat(couchdb-java-client): Migrating from ektorp to cloudant

### LINKED ISSUE #1148: Migrating from Ektorp to Cloudant Java Client
[//]: # (Copyright Siemens AG, 2021. Part of the SW360 Portal Project)
[//]: # (This program and the accompanying materials are made)
[//]: # (available under the terms of the Eclipse Public License 2.0)
[//]: # (which is available at https://www.eclipse.org/legal/epl-2.0/)
[//]: # (SPDX-License-Identifier: EPL-2.0)

### Description

Currently we are using Ektorp as a java client for CouchDB. However as we have observed earlier, it seems to be not fully compatible with Apache CouchDB 3.1.1. It would be good to migrate to some other library like [java-cloudant](https://github.com/cloudant/java-cloudant) which would be helpful in the long run.

### PR DESCRIPTION
**Pull request by smrutis1**
[//]: # (Copyright Bosch.IO GmbH 2020)
[//]: # (This program and the accompanying materials are made)
[//]: # (available under the terms of the Eclipse Public License 2.0)
[//]: # (which is available at https://www.eclipse.org/legal/epl-2.0/)
[//]: # (SPDX-License-Identifier: EPL-2.0)

> Please provide a summary of your changes here.
> * Which issue is this pull request belonging to and how is it solving it? (*#1148*)
> * Did you add or update any new dependencies that are required for your change? - Yes

Have added a new library(cloudant java client) instead of ektorp for DB operations.

Issue: Closes #1148 

### How To Test?
Need to 
> Have you implemented any additional tests? - No

### Checklist
Must:
- [ ] All related issues are referenced in commit messages and in PR

Signed-off-by: Smruti Prakash Sahoo <smruti.sahoo@siemens.com>

### CODE REVIEW

**Comment by mcjaeger** [comment]
test run cleanly on vagrant (Java 11 based, Ubuntu 20, Couchdb 311)

File: libraries/lib-datahandler/src/main/java/org/eclipse/sw360/datahandler/common/DatabaseSettings.java [hunk 1]
```diff
@@ -70,9 +76,31 @@
...
+        try {
+            clientBuilder = ClientBuilder.url(new URL(COUCH_DB_URL)).gsonBuilder(gson);
+            if (!"".equals(COUCH_DB_USERNAME)) {
+                clientBuilder.username(COUCH_DB_USERNAME);
+            }
+            if (!"".equals(COUCH_DB_PASSWORD)) {
+                clientBuilder.password(COUCH_DB_PASSWORD);
+            }
+        } catch (MalformedURLException e) {
+
```
Comment (JaideepPalit) [inline libraries/lib-datahandler/src/main/java/org/eclipse/sw360/datahandler/common/DatabaseSettings.java:98]: Add log.error()

File: libraries/lib-datahandler/src/main/java/org/eclipse/sw360/datahandler/couchdb/AttachmentConnector.java [hunk 2]
```diff
@@ -9,9 +9,11 @@
...
 package org.eclipse.sw360.datahandler.couchdb;
 
+import com.cloudant.client.api.CloudantClient;
 import com.google.common.collect.Sets;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.eclipse.sw360.datahandler.cloudantclient.DatabaseConnectorCloudant;
 import org.eclipse.sw360.datahandler.common.Duration;
 import org.eclipse.sw360.datahandler.thrift.SW360Exception;
 import org.eclipse.sw360.datahandler.thrift.attachments.Attachment;
```
Comment (JaideepPalit) [inline libraries/lib-datahandler/src/main/java/org/eclipse/sw360/datahandler/couchdb/AttachmentConnector.java:19]: Remove used imports
```
import org.ektorp.http.HttpClient;
import java.util.HashSet;
```

File: libraries/lib-datahandler/src/test/java/org/eclipse/sw360/datahandler/common/DatabaseSettingsTest.java [hunk 3]
```diff
@@ -59,6 +65,30 @@
...
+        try {
+            clientBuilder = ClientBuilder.url(new URL(COUCH_DB_URL)).gsonBuilder(gson);
+            if (!"".equals(COUCH_DB_USERNAME)) {
+                clientBuilder.username(COUCH_DB_USERNAME);
+            }
+            if (!"".equals(COUCH_DB_PASSWORD)) {
+                clientBuilder.password(COUCH_DB_PASSWORD);
+            }
+        } catch (MalformedURLException e) {
+
```
Comment (JaideepPalit) [inline libraries/lib-datahandler/src/test/java/org/eclipse/sw360/datahandler/common/DatabaseSettingsTest.java:87]: Good to have log.error()

File: libraries/lib-datahandler/src/test/java/org/eclipse/sw360/datahandler/couchdb/AttachmentStreamConnectorTest.java [hunk 4]
```diff
@@ -10,6 +10,8 @@
 package org.eclipse.sw360.datahandler.couchdb;
 
 import com.fasterxml.jackson.annotation.JsonAutoDetect;
```
Comment (JaideepPalit) [inline libraries/lib-datahandler/src/test/java/org/eclipse/sw360/datahandler/couchdb/AttachmentStreamConnectorTest.java:12]: remove unused import

**Review by JaideepPalit — CHANGES_REQUESTED** [review]
Reviewed the PR.
Kindly have a look at the review comments.

**Comment by ravi110336** [comment]
Tested the UI and it is working fine for most of the cases but found some below issue, Please check:

Issue 1:
    Add release to  a component and while fetching the release details getting the error message " Error:Error fetching release from backend."
image

Issue 2:
    Similar issue from  Admin->DataBaseSanitation->Releases with more than one source attachment:
     getting the error message " Error:Error fetching release from backend."

Issue 3:
    ECC-> click on release.  getting the error message " Error:Error fetching release from backend."

Observation 1: (Not sure if this is an issue )
Go to Requests -> click on any request -> blank result no output.
image

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/AttachmentDatabaseHandler.java [hunk 5]
```diff
@@ -147,8 +146,8 @@ public AttachmentUsage makeAttachmentUsage(AttachmentUsage attachmentUsage) {
 
     public void makeAttachmentUsages(List<AttachmentUsage> attachmentUsages) throws TException {
         List<AttachmentUsage> sanitizedUsages = distinctAttachmentUsages(attachmentUsages);
-        List<DocumentOperationResult> results = attachmentUsageRepository.executeBulk(sanitizedUsages);
-        if (!results.isEmpty()) {
+        List<Response> results = attachmentUsageRepository.executeBulk(sanitizedUsages);
+        if (CommonUtils.isNotEmpty(sanitizedUsages) && results.isEmpty()) {
             throw new SW360Exception("Some of the usage documents could not be created: " + results);
```
[the remark this answers is not in the document]
Comment (smrutis1) [reply backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/AttachmentDatabaseHandler.java:154]: Have validated the response/results before exception

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/AttachmentDatabaseHandler.java [hunk 6]
```diff
@@ -185,20 +184,20 @@ public AttachmentUsage updateAttachmentUsage(AttachmentUsage attachmentUsage) {
     }
 
     public void updateAttachmentUsages(List<AttachmentUsage> attachmentUsages) throws TException {
-        List<DocumentOperationResult> results = attachmentUsageRepository.executeBulk(attachmentUsages);
+        List<Response> results = attachmentUsageRepository.executeBulk(attachmentUsages);
```
[the remark this answers is not in the document]
Comment (smrutis1) [reply backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/AttachmentDatabaseHandler.java:190]: Have added the validation

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ProjectVulnerabilityRatingRepository.java [hunk 7]
```diff
@@ -34,17 +37,38 @@
...
+        initStandardDesignDocument(views, db);
     }
 
     @View(name = "projectVulnerabilityRating", map = PROJECT_VULNERABILITY_LINK)
     public List<ProjectVulnerabilityRating> getProjectVulnerabilityRating(String projectId) {
         List<ProjectVulnerabilityRating> queryResults = queryByPrefix("projectVulnerabilityRating", projectId);
 
         if (queryResults.size() > 1) {
-            log.error("More than one projectVulnerabilityRating found for project with id " + projectId);
+//            log.error("More than one projectVulnerabilityRating found for project with id " + projectId);
```
[the remark this answers is not in the document]
Comment (smrutis1) [reply backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ProjectVulnerabilityRatingRepository.java]: Have added the logger

File: backend/src/src-licenses/src/main/java/org/eclipse/sw360/licenses/db/LicenseDatabaseHandler.java [hunk 8]
```diff
@@ -478,7 +482,7 @@ public RequestStatus updateLicenseFromAdditionsAndDeletions(License licenseAddit
         if (!PermissionUtils.isUserAtLeast(UserGroup.CLEARING_ADMIN, user)){
             return null;
         }
-        final List<DocumentOperationResult> documentOperationResults = licenseTypeRepository.executeBulk(licenseTypes);
+        final List<Response> documentOperationResults = licenseTypeRepository.executeBulk(licenseTypes);
```
[the remark this answers is not in the document]
Comment (smrutis1) [reply backend/src/src-licenses/src/main/java/org/eclipse/sw360/licenses/db/LicenseDatabaseHandler.java]: Have added the validation

File: backend/src/src-moderation/src/main/java/org/eclipse/sw360/moderation/db/ModerationRequestRepository.java [hunk 9]
```diff
@@ -48,10 +52,14 @@
...
-        initStandardDesignDocument();
+        Map<String, MapReduce> views = new HashMap<String, MapReduce>();
+        views.put("all", createMapReduce(ALL, null));
+        views.put("documents", createMapReduce(DOCUMENTS_VIEW, null));
+        views.put("moderators", createMapReduce(MODERATORS_VIEW, null));
+        views.put("users", createMapReduce(USERS_VIEW, null));
+        initStandardDesignDocument(views, db);
     }
 
     @View(name = "documents", map = DOCUMENTS_VIEW)
```
[the remark this answers is not in the document]
Comment (smrutis1) [reply backend/src/src-moderation/src/main/java/org/eclipse/sw360/moderation/db/ModerationRequestRepository.java]: Removed it

File: .travis.yml [hunk 10]
```diff
@@ -31,8 +32,39 @@ cache:
...
+  - echo "couchdb.password = ${COUCHDB_ADMIN_PWD}" | sudo tee -a /etc/sw360/couchdb-test.properties
+  - echo "couchdb.database = sw360_test_db" | sudo tee -a /etc/sw360/couchdb-test.properties
+  - echo "couchdb.usersdb = sw360_test_users" | sudo tee -a /etc/sw360/couchdb-test.properties
+  - echo "couchdb.attachments = sw360_test_attachments" | sudo tee -a /etc/sw360/couchdb-test.properties
+  - echo "couchdb.fossologyKeys = sw360_test_fossologykeys" | sudo tee -a /etc/sw360/couchdb-test.properties
+  - echo "couchdb.vulnerability_management = sw360_test_vm" | sudo tee -a /etc/sw360/couchdb-test.properties
+  - echo "couchdb.change_logs = sw360_test_changelogs" | sudo tee -a /etc/sw360/couchdb-test.properties
+  - echo "couchdb.config = sw360_test_config" | sudo tee -a /etc/sw360/couchdb-test.properties
+
+    # Create couchdb.properties in /etc/sw360
```
[the remark this answers is not in the document]
Comment (smrutis1) [reply .travis.yml:51]: In `DatabaseHandlerUtil.java` the changelog repository is being created in a static way. Hence the initialization of it fails without credentials and DB name in prop. file. (Ref1 https://github.com/eclipse/sw360/blob/4bf2932ba060384f5cd6975ef0e1fac78692dca8/backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/DatabaseHandlerUtil.java#L94   Ref: https://github.com/eclipse/sw360/blob/4bf2932ba060384f5cd6975ef0e1fac78692dca8/backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/DatabaseHandlerUtil.java#L316)

May be the construction of DatabaseHandlerUtil needs to changed to handle it properly. Let me know if it should be fixed here.

**Comment by smrutis1** [comment]
@JaideepPalit , Thanks for the review. Have incorporated most of them. Kindly have a look and let me know. @ravi110336 Thanks for the test. Have fixed the mentioned issues however the observation regarding "Request" tab, I am not able to reproduce it.

File: .travis.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (smrutis1) [reply .travis.yml]: From discussion, have handled it here.

**Comment by smrutis1** [comment]
@ravi110336 , From discussion, have rectified the issue with moderation request. Kindly have a look

**Comment by ravi110336** [comment]
Tested the moderation request and attachment download working fine.

**Comment by mcjaeger** [comment]
tested also with vagrant: looks good

**Review by JaideepPalit — CHANGES_REQUESTED** [review]
**Re-reviewed the PR.**
Kindly have a look at some very minor comments.

**Note-** 
The travis-ci logs gives lot of error messages during test, which are not coming in other branches.
https://travis-ci.org/github/eclipse/sw360/jobs/768972323#L2046
```
Apr 30, 2021 6:56:28 AM com.cloudant.client.org.lightcouch.CouchDbClient createDB
INFO: Created Database: 'sw360_test_db'
2021-04-30 06:56:28 ERROR DatabaseConnectorCloudant:138 - Error fetching document of type DesignDocument with id _design/AttachmentUsage : 404 Object Not Found at http://localhost:5984/sw360_test_db/_design/AttachmentUsage. Error: not_found. Reason: missing.
2021-04-30 06:56:28 ERROR DatabaseConnectorCloudant:138 - Error fetching document of type DesignDocument with id _design/Attachment : 404 Object Not Found at http://localhost:5984/sw360_test_db/_design/Attachment. Error: not_found. Reason: missing.
2021-04-30 06:56:28 ERROR DatabaseConnectorCloudant:138 - Error fetching document of type DesignDocument with id _design/Source : 404 Object Not Found at http://localhost:5984/sw360_test_db/_design/Source. Error: not_found. Reason: missing.
2021-04-30 06:56:28 ERROR DatabaseConnectorCloudant:138 - Error fetching document of type DesignDocument with id _design/Vendor : 404 Object Not Found at http://localhost:5984/sw360_test_db/_design/Vendor. Error: not_found. Reason: missing.
2021-04-30 06:56:28 ERROR DatabaseConnectorCloudant:138 - Error fetching document of type DesignDocument with id _design/Release : 404 Object Not Found at http://localhost:5984/sw360_test_db/_design/Release. Error: not_found. Reason: missing.
2021-04-30 06:56:29 ERROR DatabaseConnectorCloudant:138 - Error fetching document of type DesignDocument with id _design/Component : 404 Object Not Found at http://localhost:5984/sw360_test_db/_design/Component. Error: not_found. Reason: missing.
2021-04-30 06:56:29 ERROR DatabaseConnectorCloudant:138 - Error fetching document of type DesignDocument with id _design/Project : 404 Object Not Found at http://localhost:5984/sw360_test_db/_design/Project. Error: not_found. Reason: missing.
2021-04-30 06:56:29 ERROR DatabaseConnectorCloudant:138 - Error fetching document of type DesignDocument with id _design/User : 404 Object Not Found at http://localhost:5984/sw360_test_db/_design/User. Error: not_found. Reason: missing.
2021-04-30 06:56:29 ERROR DatabaseConnectorCloudant:138 - Error fetching document of type DesignDocument with id _design/ProjectVulnerabilityRating : 404 Object Not Found at http://localhost:5984/sw360_test_db/_design/ProjectVulnerabilityRating. Error: not_found. Reason: missing.
2021-04-30 06:56:29 ERROR DatabaseConnectorCloudant:138 - Error fetching document of type DesignDocument with id 
```
Seems like , using cloudant - design docs are not getting generated. 
It should also be tested on a clean system with no DBs or Design docs

**Comment by ravi110336** [comment]
I have tested the branch after deleting all the DBs and design docs in my local system, all the DBs are getting created with the design docs successfully.

**Comment by smrutis1** [comment]
@JaideepPalit Thanks for review. Have incorporated changes.
> **Note-**
> The travis-ci logs gives lot of error messages during test, which are not coming in other branches.
> https://travis-ci.org/github/eclipse/sw360/jobs/768972323#L2046

(from discussion): This is due to added logger message, it's ok to keep it.
* Have added the paginated view for components to mitigate loading issue

**Review by JaideepPalit — CHANGES_REQUESTED** [review]
Reviewed the PR.
Kindly have a look at some review comments.
Also seems to be some issue with 
 1. sorting based on vendor and main licenses 
     - case-sensitivity
     - Multivalue list
sort1

 2. There seems to some issue with pagination. Based on columns sorted like vendor/license/name - total result seems to vary
comp_name
main_lic

**Comment by smrutis1** [comment]
> Reviewed the PR.
> Kindly have a look at some review comments.
> Also seems to be some issue with
> 
> 1. sorting based on vendor and main licenses
>    
>    * case-sensitivity
>    * Multivalue list\

(from discussion) Raised an issue #1199 
>     
> 2. There seems to some issue with pagination. Based on columns sorted like vendor/license/name - total result seems to vary
>    comp_name
>    main_lic

Fixed it.

**Review by JaideepPalit — APPROVED** [review]
Code looks fine to me.

**Comment by mcjaeger** [comment]
has been tested on our staging server, vagrant of course and again staging server....

**Pull request merged**